subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/IsolatingIncrementalAnnotationProcessingIntegrationTest.groovy

```diff
@@ -275,6 +306,133 @@ class IsolatingIncrementalAnnotationProcessingIntegrationTest extends AbstractIn
         outputContains("Full recompilation is required because the generated type 'ServiceRegistry' must have exactly one originating element, but had 2.")
     }
 
+    def "processors cannot provide multiple originating elements for resources"() {
+        given:
+        def proc = new ServiceRegistryProcessorFixture()
+        proc.writeResources = true
+        withProcessor(proc.withDeclaredType(IncrementalAnnotationProcessorType.ISOLATING))
+        def a = java "@Service class A {}"
+        java "@Service class B {}"
+
+        outputs.snapshot { succeeds "compileJava" }
+
+        when:
+        a.text = "@Service class A { void foo() {} }"
+
+        then:
+        succeeds "compileJava", "--info"
+
+        and:
+        outputContains("Full recompilation is required because the generated resource 'ServiceRegistryResource.txt in CLASS_OUTPUT' must have exactly one originating element, but had 2.")
+    }
+
+    def "processors can generate identical resources in different locations"() {
+        given:
+        // Have to configure a native header output directory otherwise there will be errors; javac NPEs when files are created in NATIVE_HEADER_OUTPUT without any location set.
+        buildFile << '''
+compileJava.options.headerOutputDirectory = file("build/headers/java/main")
+'''
+        def locations = [StandardLocation.SOURCE_OUTPUT.toString(), StandardLocation.NATIVE_HEADER_OUTPUT.toString(), StandardLocation.CLASS_OUTPUT.toString()]
+        withProcessor(new ResourceGeneratingProcessorFixture().withOutputLocations(locations).withDeclaredType(IncrementalAnnotationProcessorType.ISOLATING))
+        def a = java "@Thing class A {}"
+        java "class Unrelated {}"
+
+        when:
+        outputs.snapshot { succeeds "compileJava" }
+
+        then:
+        file("build/generated/sources/annotationProcessor/java/main/A.txt").exists()
+        file("build/headers/java/main/A.txt").exists()
+        file("build/classes/java/main/A.txt").exists()
+
+        when:
+        a.delete()
+        succeeds "compileJava"
+
+        then: "they all get cleaned"
+        outputs.deletedClasses("A")
+        !file("build/generated/sources/annotationProcessor/java/main/A.txt").exists()
+        !file("build/headers/java/main/A.txt").exists()
+        !file("build/classes/java/main/A.txt").exists()
+    }
+
+    def "SOURCE_OUTPUT resources get put into CLASS_OUTPUT if a location is not configured"() {
```

Comment:
This is testing javac, not Gradle, so doesn't need to be here.

Reply:
I thought of this test (and some of the other ones) covering the bit in IncrementalCompilationInitializer above that we discussed.  Because that code is actually relying on javac's behavior, shouldn't we test javac's behavior?

Reply:
I wouldn't, it's not just javac that does that. Alternatively you can go with the disabling incremental route if you feel that would be a safer strategy.